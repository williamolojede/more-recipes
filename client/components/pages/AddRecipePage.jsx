import React from 'react';

import ManageRecipes from '../shared/ManageRecipes';
import AddRecipeForm from '../shared/AddRecipeForm';

const AddRecipePage = props => (
  <ManageRecipes>
    <AddRecipeForm isEditMode={false} {...props} />
  </ManageRecipes>
);
export default AddRecipePage;
