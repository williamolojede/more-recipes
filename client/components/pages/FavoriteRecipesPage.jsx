import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import ReactPaginate from 'react-paginate';
import swal from 'sweetalert';


import decodeToken from '../../utils/decodeToken';

import ManageRecipes from '../shared/ManageRecipes';
import Preloader from '../shared/Preloader';
import RecipeCardShort from '../shared/RecipeCard/RecipeCardShort';

import { fetchFavoriteRecipes, favorite } from '../../actions/recipe';

class FavoriteRecipesPage extends Component {
  constructor(props) {
    super(props);
    this.state = {
      recipes: [],
      currentPage: 1,
      limit: 12,
      last: 1,
    };

    this.payload = {
      userId: decodeToken().id,
      page: this.state.currentPage,
      limit: this.state.limit
    };
  }

  componentWillMount() {
    const payload = {
      userId: decodeToken().id,
      page: this.state.currentPage,
      limit: this.state.limit
    };
    this.props.dispatch(fetchFavoriteRecipes(payload));
  }

  componentWillReceiveProps(newProps) {
    if (newProps.recipes) {
      const {
        recipes,
        pagination
      } = newProps;
      this.setState({
        recipes,
        last: pagination.last
      });
    }
  }

  componentDidUpdate(prevProps, prevState) {
    if (prevState.currentPage !== this.state.currentPage) {
      const payload = {
        userId: decodeToken().id,
        page: this.state.currentPage,
        limit: this.state.limit
      };
      this.props.dispatch(fetchFavoriteRecipes(payload));
    }
  }

  handlePageClick = ({ selected }) => {
    this.setState({
      currentPage: selected + 1
    });
  }

  removeRecipe = (id) => {
    swal({
      title: 'Remove recipe from your favorite List?',
      icon: 'warning',
      buttons: true,
      dangerMode: true,
    })
      .then((willDelete) => {
        if (willDelete) {
          return this.props.dispatch(favorite(id));
        }
        return willDelete;
      })
      .then((willDelete) => {
        if (!willDelete) return;
        const payload = {
          userId: decodeToken().id,
          page: (this.state.recipes.length === 1 && this.state.last !== 1) ?
            this.state.currentPage - 1 : this.state.currentPage,
          limit: this.state.limit
        };
        this.props.dispatch(fetchFavoriteRecipes(payload));
      });
  }

  renderBody = () => {
    const { isFetching } = this.props;

    if (isFetching) {
      return (<Preloader />);
    } else if (!isFetching && this.state.recipes.length === 0) {
      return (
        <div className="no-favorites">
          <span>Welcome to your favorites!</span>
          <span>Tap <i className="mdi mdi-heart" /> to save any recipe you like,</span>
          <span>and it&apos;ll show up here.</span>
        </div>
      );
    }
    return (
      <ul className="row recipes-list">
        {
          this.state.recipes.map(({ id, name }) => (<RecipeCardShort
            id={id}
            name={name}
            removeRecipe={this.removeRecipe}
            key={id}
            type="favorite"
          />))
        }
      </ul>
    );
  }

  render() {
    return (
      <ManageRecipes>
        <div className="container page-manage__recipes">

          {
            this.renderBody()
          }
          {
            (this.state.last > 1 && this.state.recipes.length !== 0) &&
            <footer>
              <ReactPaginate
                previousLabel={<i className="material-icons">chevron_left</i>}
                nextLabel={<i className="material-icons">chevron_right</i>}
                pageCount={this.state.last}
                marginPagesDisplayed={2}
                pageClassName="waves-effect"
                pageRangeDisplayed={4}
                onPageChange={this.handlePageClick}
                activeClassName="active"
                containerClassName={`pagination ${this.props.isFetching ? 'is-fetching' : ''}`}
              />
            </footer>
          }

        </div>
      </ManageRecipes>
    );
  }
}

FavoriteRecipesPage.propTypes = {
  dispatch: PropTypes.func.isRequired,
  isFetching: PropTypes.bool.isRequired,
};

const mapStateToProps = ({
  favoriteRecipes: { recipes, pagination },
  isFetching,
}) => ({
  recipes,
  pagination,
  isFetching
});

export { FavoriteRecipesPage as PureFavoriteRecipesPage };
export default connect(mapStateToProps)(FavoriteRecipesPage);
