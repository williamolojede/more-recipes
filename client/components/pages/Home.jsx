import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import ReactPaginate from 'react-paginate';

import { fetchTopRecipes } from '../../actions/recipe';

import TopRatedRecipeList from '../shared/TopRatedRecipeList';
import Preloader from '../shared/Preloader';
import ConnectedSiteNav from '../shared/SiteNav';
import SiteFooter from '../shared/SiteFooter';
import Search from '../shared/Search';

import { currentUserPropTypes } from '../../config/proptypes';

class Home extends Component {
  constructor(props) {
    super(props);
    this.state = {
      recipes: [],
      currentPage: 1,
      limit: 6,
      last: 1,
      isSearchActive: false
    };
  }

  componentDidMount() {
    this.props.dispatch(fetchTopRecipes(this.state.currentPage, this.state.limit));
  }

  componentWillReceiveProps(newProps) {
    if (newProps.recipes) {
      const {
        recipes,
        pagination
      } = newProps;
      this.setState({
        recipes,
        last: pagination.last
      });
    }
  }

  componentDidUpdate(prevProps, prevState) {
    if (prevState.currentPage !== this.state.currentPage) {
      this.props.dispatch(fetchTopRecipes(this.state.currentPage, this.state.limit));
    }
  }

  toggleSearch =() => {
    this.setState({
      isSearchActive: !this.state.isSearchActive
    });
  }

  handlePageClick = ({ selected }) => {
    this.setState({
      currentPage: selected + 1
    });
  }

  renderBody = () => {
    const { isFetching } = this.props;

    if (isFetching) {
      return (<Preloader />);
    } else if (!isFetching && this.state.recipes.length === 0) {
      return (
        <p className="no-recipes">
          Sorry no recipes present at the moment
          <span role="img" aria-label="sad">üò¢  </span>, please try again later
          <span role="img" aria-label="please">üôè</span>
        </p>
      );
    }
    return (<TopRatedRecipeList recipes={this.state.recipes} />);
  }

  render() {
    return (
      <div className="page page__home">
        <header className="site-header">
          <ConnectedSiteNav currentUser={this.props.currentUser} />
          <section className="site-header__hero">
            <h2>Delicious recipes, just a search away...</h2>
            <input
              className="site-header__hero--input"
              type="text"
              onFocus={this.toggleSearch}
              placeholder="what would you like to cook?"
            />
          </section>
        </header>
        <main>
          <div className="container">
            <section className="page__home--top-rated">
              <header>
                <h1>Top Rated</h1>
              </header>
              {
                this.renderBody()
              }
              <footer>
                <ReactPaginate
                  previousLabel={<i className="material-icons">chevron_left</i>}
                  nextLabel={<i className="material-icons">chevron_right</i>}
                  pageCount={this.state.last}
                  marginPagesDisplayed={2}
                  pageClassName="waves-effect"
                  pageRangeDisplayed={4}
                  onPageChange={this.handlePageClick}
                  activeClassName="active"
                  containerClassName="pagination"
                />
              </footer>


            </section>
          </div>
        </main>
        {
          this.state.isSearchActive && <Search toggleSearch={this.toggleSearch} />
        }
        <SiteFooter />
      </div>
    );
  }
}

Home.propTypes = {
  dispatch: PropTypes.func.isRequired,
  isFetching: PropTypes.bool.isRequired,

  ...currentUserPropTypes
};

const mapStateToProps = ({
  topRecipes: { recipes, pagination },
  isFetching,
  auth: { currentUser }
}) => ({
  recipes,
  pagination,
  currentUser,
  isFetching
});

export { Home as PureHome };
export default connect(mapStateToProps)(Home);
