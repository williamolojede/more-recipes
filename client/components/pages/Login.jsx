import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Redirect, Link } from 'react-router-dom';

import { authPropTypes } from '../../config/proptypes';

import Footer from '../shared/Footer.jsx';
import AuthForm from '../shared/AuthForm.jsx';
import Preloader from '../shared/Preloader.jsx';
import ErrorDisplay from '../shared/ErrorDisplay.jsx';

import loginUser from '../../actions/loginUser';
import { removeAuthErrorMessage } from '../../actions/auth';


class Login extends Component {
  componentWillMount() {
    if (this.props.errorMessage) {
      this.props.dispatch(removeAuthErrorMessage());
    }
  }

  /**
   * @param {object} formData -
   * @returns {undefined} - return nothing
   */
  userLogin = (formData) => {
    this.props.dispatch(loginUser(formData));
  }

  render() {
    const {
      isFetching,
      errorMessage,
      isAuthenticated,
      location
    } = this.props;
    const { from } = location.state || { from: { pathname: '/' } };
    if (isAuthenticated) {
      return (
        <Redirect to={from} />
      );
    }

    return (
      <div className="page page__login">
        <header className="header center">
          <a href="/" className="white-text brand-logo">MoreRecipes</a>
        </header>

        <main>
          <div className="container">
            <div className="row">
              <div className="login-card card col s11 m6">
                <div className="card-content">
                  {
                    errorMessage ? <ErrorDisplay message={errorMessage} /> : null
                  }
                  <h1 className="card-title center">Log in to MoreRecipes</h1>
                  <AuthForm type="login" authFormSubmit={this.userLogin} />
                </div>
              </div>
            </div>
            <p className="white-text center">
            Donâ€™t have an account?
              <Link to="/signup" className="white-text"> Signup now</Link>
            </p>
            {
              isFetching ? <Preloader /> : null
            }
          </div>
        </main>

        <Footer />
      </div>
    );
  }
}

Login.defaultProps = {
  // the error message should not be fined by default
  errorMessage: undefined
};

Login.propTypes = authPropTypes;

export const mapStateToProps = ({ auth, isFetching }) => {
  const { isAuthenticated, errorMessage } = auth;
  return {
    isAuthenticated,
    errorMessage,
    isFetching
  };
};

export { Login as PureLogin };
export default connect(mapStateToProps)(Login);
