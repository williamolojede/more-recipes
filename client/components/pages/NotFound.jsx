import React from 'react';
import { Link } from 'react-router-dom';

const NotFound = () => (
  <div className="page page__not-found">
    <main>
      <h1>404</h1>
      <p>Sorry, the web page your are looking for is missing
        <span role="img" aria-label="cry">ðŸ˜­</span>
      </p>
      <Link to="/">Go back to the home page</Link>
    </main>
  </div>
);


export default NotFound;
