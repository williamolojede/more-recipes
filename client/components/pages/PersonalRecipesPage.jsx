import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import ReactPaginate from 'react-paginate';
import swal from 'sweetalert';

import decodeToken from '../../utils/decodeToken';

import ManageRecipes from '../shared/ManageRecipes';
import Preloader from '../shared/Preloader';
import RecipeCardShort from '../shared/RecipeCard/RecipeCardShort';

import { fetchPersonalRecipes, deleteRecipe } from '../../actions/recipe';

class PersonalRecipesPage extends Component {
  constructor(props) {
    super(props);
    this.state = {
      recipes: [],
      currentPage: 1,
      limit: 12,
      last: 1,
    };

    this.payload = {
      userId: decodeToken().id,
      page: this.state.currentPage,
      limit: this.state.limit
    };
  }

  componentWillMount() {
    const payload = {
      userId: decodeToken().id,
      page: this.state.currentPage,
      limit: this.state.limit
    };
    // console.log(this.payload);
    this.props.dispatch(fetchPersonalRecipes(payload));
  }

  componentWillReceiveProps(newProps) {
    if (newProps.recipes) {
      const {
        recipes,
        pagination
      } = newProps;
      this.setState({
        recipes,
        last: pagination.last
      });
    }
  }

  componentDidUpdate(prevProps, prevState) {
    if (prevState.currentPage !== this.state.currentPage) {
      const payload = {
        userId: decodeToken().id,
        page: this.state.currentPage,
        limit: this.state.limit
      };
      this.props.dispatch(fetchPersonalRecipes(payload));
    }
  }

  handlePageClick = ({ selected }) => {
    this.setState({
      currentPage: selected + 1
    });
  }

  removeRecipe = (id) => {
    swal({
      title: 'Are you sure?',
      text: 'Once deleted, you will not be able to recover this recipe!',
      icon: 'warning',
      buttons: true,
      dangerMode: true,
    })
      .then((willDelete) => {
        if (willDelete) {
          return this.props.dispatch(deleteRecipe(id));
        }
        return willDelete;
      })
      .then((willDelete) => {
        if (!willDelete) return;
        const payload = {
          userId: decodeToken().id,
          page: (this.state.recipes.length === 1 && this.state.last !== 1) ?
            this.state.currentPage - 1 : this.state.currentPage,
          limit: this.state.limit
        };
        this.props.dispatch(fetchPersonalRecipes(payload));
      });
  }

  renderBody = () => {
    const { isFetching } = this.props;

    if (isFetching) {
      return (<Preloader />);
    } else if (!isFetching && this.state.recipes.length === 0) {
      return (
        <div className="no-favorites">
          <span>You don&apos;t have any personal recipes yet.</span>
          <span>Share your own culinary creations! Add a recipe</span>
        </div>
      );
    }
    return (
      <ul className="row recipes-list">
        {
          this.state.recipes.map(({ id, name }) => (<RecipeCardShort
            id={id}
            name={name}
            key={id}
            removeRecipe={this.removeRecipe}
            type="personal"
          />))
        }
      </ul>
    );
  }

  render() {
    return (
      <ManageRecipes>
        <div className="container page-manage__recipes">

          {
            this.renderBody()
          }
          {
            (this.state.last > 1 && this.state.recipes.length !== 0) &&
            <footer>
              <ReactPaginate
                previousLabel={<i className="material-icons">chevron_left</i>}
                nextLabel={<i className="material-icons">chevron_right</i>}
                pageCount={this.state.last}
                marginPagesDisplayed={2}
                pageClassName="waves-effect"
                pageRangeDisplayed={4}
                onPageChange={this.handlePageClick}
                activeClassName="active"
                containerClassName={`pagination ${this.props.isFetching ? 'is-fetching' : ''}`}
              />
            </footer>
          }

        </div>
      </ManageRecipes>
    );
  }
}

PersonalRecipesPage.propTypes = {
  dispatch: PropTypes.func.isRequired,
  isFetching: PropTypes.bool.isRequired,
};

const mapStateToProps = ({
  personalRecipes: { recipes, pagination },
  isFetching,
}) => ({
  recipes,
  pagination,
  isFetching
});

export { PersonalRecipesPage as PurePersonalRecipesPage };
export default connect(mapStateToProps)(PersonalRecipesPage);
