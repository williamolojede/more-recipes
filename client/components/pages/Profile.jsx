import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { currentUserPropTypes, notificationPropTypes } from '../../config/proptypes';
import { updateUserProfile } from '../../actions/userProfile';
import defaultConfig from '../../config/default';
import imageUpload from '../../utils/imageUpload';

import ConnectedSiteNav from '../shared/SiteNav.jsx';
import Preloader from '../shared/Preloader.jsx';
import SiteFooter from '../shared/SiteFooter';
import ImageUpload from '../shared/ImageUpload';
import Notification from '../shared/Notification';
import SubmitButton from '../shared/SubmitButton';

class Profile extends Component {
  constructor(props) {
    super(props);

    this.state = {
      user: this.props.currentUser,
      // pending, started, failed, completed
      imageUploadProgress: 'pending',
      submitLoading: false
    };
  }

  componentWillReceiveProps(newProps) {
    if (newProps.currentUser.id !== 0) {
      this.setState({ user: newProps.currentUser });
    }
  }

  imageUploadHandler = (imageInput) => {
    imageUpload.call(
      this,
      imageInput,
      `users/${this.props.currentUser.id}--${Date.now()}`,
      (downloadURL) => {
        const user = Object.assign({}, this.state.user);
        user.imgUrl = downloadURL;
        imageInput.disabled = false;
        this.setState({ user, imageUploadProgress: 'completed' });
      }
    );
  }

  handleUpdateProfile = () => {
    this.setState({ submitLoading: true });
    this.props.dispatch(updateUserProfile(this.state.user))
      .then(() => this.setState({ submitLoading: false }));
  }

  handleOnChange = (event, type) => {
    const user = Object.assign({}, this.state.user);
    user[type] = event.target.value;
    this.setState({ user });
  }

  render() {
    const { currentUser } = this.props;

    const {
      fullname,
      email,
    } = this.state.user;

    return (
      this.props.isFetching ?
        <Preloader />
        :
        <div className="page page__profile">
          <header className="site-header">
            <ConnectedSiteNav currentUser={currentUser} />
          </header>
          <main>
            <div className="container z-depth-1">
              <section className="profile">
                <h3 className="profile__title">Profile</h3>
                <div className="profile__body">
                  <ImageUpload
                    imgUrl={this.state.user.imgUrl}
                    className="profile__image"
                    altText={fullname}
                    imageUploadHandler={this.imageUploadHandler}
                    imageUploadProgress={this.state.imageUploadProgress}
                    defaultImgUrl={defaultConfig.userImgUrl}
                  />
                  <div className="profile__details">
                    <label htmlFor="fullname">
                      Name
                      <input
                        type="text"
                        id="fullname"
                        value={fullname}
                        onChange={e => this.handleOnChange(e, 'fullname')}
                      />
                    </label>
                    <label htmlFor="email">
                      Email
                      <input
                        type="text"
                        id="email"
                        value={email}
                        onChange={e => this.handleOnChange(e, 'email')}
                      />
                    </label>
                    <SubmitButton
                      submitText="Save Changes"
                      handleClick={this.handleUpdateProfile}
                      submitLoading={this.state.submitLoading}
                    />
                  </div>
                </div>

              </section>
            </div>
          </main>
          <Notification
            dispatch={this.props.dispatch}
            notification={this.props.notification}
          />
          <SiteFooter />
        </div>
    );
  }
}

Profile.propTypes = {
  ...currentUserPropTypes,
  isFetching: PropTypes.bool.isRequired,
  dispatch: PropTypes.func.isRequired,
  ...notificationPropTypes
};

export const mapStateToProps = ({
  isFetching,
  auth: { currentUser },
  notification
}) => ({
  currentUser,
  isFetching,
  notification
});

export { Profile as PureProfile };
export default connect(mapStateToProps)(Profile);
