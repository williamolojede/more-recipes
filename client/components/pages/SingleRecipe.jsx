/* eslint camelcase: 0 */

import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';


import {
  fetchSingleRecipe,
  vote,
  favorite,
  resetSingleRecipe
} from '../../actions/recipe';


import Preloader from '../shared/Preloader.jsx';
import ConnectedSiteNav from '../shared/SiteNav.jsx';
import SiteFooter from '../shared/SiteFooter.jsx';
import IngredientsList from '../shared/IngredientsList.jsx';
import InstructionsList from '../shared/InstructionsList.jsx';
import Reviews from '../shared/Reviews';
import RecipeStats from '../shared/RecipeStats.jsx';
import Notification from '../shared/Notification.jsx';

import { recipePropTypes, currentUserPropTypes } from '../../config/proptypes';

class SingleRecipe extends Component {
  componentDidMount() {
    this.props.dispatch(fetchSingleRecipe(this.props.match.params.id));
  }

  componentWillUnmount() {
    this.props.dispatch(resetSingleRecipe());
  }

  vote = (dir) => {
    this.props.dispatch(vote(dir, this.props.recipe.id));
  }

  favorite = () => {
    this.props.dispatch(favorite(this.props.recipe.id));
  }

  renderBody = () => {
    const { isFetching } = this.props;
    const {
      name,
      description,
      img_url,
      ingredients,
      instructions,
      upVoteCount,
      downVoteCount,
      favoriteCount,
      viewCount
    } = this.props.recipe;

    const stats = {
      upVoteCount,
      downVoteCount,
      favoriteCount,
      viewCount
    };

    // the default value of recipe.name => ''
    // which means the recipe hasnt been received yet
    if (isFetching && this.props.recipe.name === '') {
      return (<Preloader />);
    } else if (this.props.errorMessage) {
      return (
        <div className="recipe-not-found">
          <h1>
            <span role="img" aria-label="sad">ðŸ˜¢</span> Recipe Not Found
          </h1>
        </div>
      );
    }

    return (
      <main>
        <div className="container">
          <div className="row">
            <div className="col s12 m5">
              <img src={img_url} alt={name} />
            </div>
            <div className="col s12 m7">
              <h4 className="recipe__name">{name}</h4>
              <p className="recipe__description">{description}</p>
              <div className="row recipe__stats">
                <RecipeStats stats={stats} vote={this.vote} favorite={this.favorite} />
                <a href="#write-review" className="write-review">
                Write Review
                </a>
              </div>
            </div>
          </div>

          <IngredientsList ingredients={ingredients} />
          <InstructionsList instructions={instructions} />
          {
            name !== '' && <Reviews recipeId={parseInt(this.props.match.params.id, 10)} />
          }
        </div>
      </main>
    );
  };

  render() {
    return (
      <div className="page page__recipe">
        <header className="site-header">
          <ConnectedSiteNav currentUser={this.props.currentUser} />
        </header>
        {
          this.renderBody()
        }
        <Notification notification={this.props.notification} dispatch={this.props.dispatch} />
        <SiteFooter />
      </div>
    );
  }
}

SingleRecipe.propTypes = {
  dispatch: PropTypes.func.isRequired,

  ...recipePropTypes,
  ...currentUserPropTypes,

  match: PropTypes.shape({
    params: PropTypes.shape({
      id: PropTypes.string.isRequired,
    })
  }).isRequired,

  notification: PropTypes.string.isRequired
};

export const mapStateToProps = ({
  singleRecipe: { recipe, errorMessage },
  isFetching,
  auth: { currentUser },
  notification
}) => ({
  recipe,
  errorMessage,
  isFetching,
  notification,
  currentUser,
});

export { SingleRecipe as PureSingleRecipe };
export default connect(mapStateToProps)(SingleRecipe);
