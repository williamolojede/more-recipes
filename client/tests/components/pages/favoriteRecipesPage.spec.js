import React from 'react';
import { PureFavoriteRecipesPage } from '../../../components/pages/FavoriteRecipesPage';
import recipesMock from '../../__mocks__/recipes.mock';

function setup() {
  const props = {
    dispatch: jest.fn(),
    isFetching: false,
    recipes: [],
    pagination: {
      last: 1,
    }
  };
  const wrapper = shallow(<PureFavoriteRecipesPage {...props} />);

  return {
    wrapper
  };
}

describe('Favorite Recipes Page', () => {
  it('should render with right amount of elements', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a preloader', () => {
    const { wrapper } = setup();
    wrapper.setProps({ isFetching: true });
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a list of recipes', () => {
    const { wrapper } = setup();
    wrapper.setProps({
      recipes: recipesMock.recipes,
      pagination: recipesMock.pagination
    });
    expect(wrapper).toMatchSnapshot();
  });

  describe('handlePageClick', () => {
    it('should update currentPage in state and trigger action to fetch recipes for the next page', () => {
      const { wrapper } = setup();
      const dispatchSpy = jest.spyOn(wrapper.instance().props, 'dispatch');
      wrapper.instance().handlePageClick({ selected: 1 });

      expect(wrapper.instance().state.currentPage).toBe(2);
      // 2 because dispatch has been called first in CWM
      expect(dispatchSpy).toHaveBeenCalledTimes(2);
    });
  });

  describe('removeRecipe', () => {
    it('should promt user to confirm delete and then dispatch an action to delete recipe', () => {
      const { wrapper } = setup();
      // const dispatchSpy = jest.spyOn(wrapper.instance().props, 'dispatch');
      wrapper.instance().removeRecipe(1);
    });
  });
});
