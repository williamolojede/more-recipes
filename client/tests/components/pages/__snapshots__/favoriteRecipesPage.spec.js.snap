// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favorite Recipes Page should render a list of recipes 1`] = `
<Connect(ManageRecipes)>
  <div
    className="container page-manage__recipes"
  >
    <ul
      className="row recipes-list"
    >
      <RecipeCardShort
        id={2}
        key="2"
        name="Jollof Rice"
        removeRecipe={[Function]}
        type="favorite"
      />
      <RecipeCardShort
        id={5}
        key="5"
        name="Chef John's Spaghetti with Red Clam Sauce dfbefadb nifek fedckltefdkletfdmlefdleff eifdanlkref"
        removeRecipe={[Function]}
        type="favorite"
      />
      <RecipeCardShort
        id={169}
        key="169"
        name="some random very delicious dish"
        removeRecipe={[Function]}
        type="favorite"
      />
      <RecipeCardShort
        id={25}
        key="25"
        name="Chef John's Spaghetti with Red Clam Sauce dfbefadb nifek fedckltefdkletfdmlefdleff eifdanlkref"
        removeRecipe={[Function]}
        type="favorite"
      />
      <RecipeCardShort
        id={91}
        key="91"
        name="Chef John's Spaghetti with Red Clam Sauce dfbefadb nifek fedckltefdkletfdmlefdleff eifdanlkref"
        removeRecipe={[Function]}
        type="favorite"
      />
      <RecipeCardShort
        id={28}
        key="28"
        name="Chef John's Spaghetti with Red Clam Sauce dfbefadb nifek fedckltefdkletfdmlefdleff eifdanlkref"
        removeRecipe={[Function]}
        type="favorite"
      />
    </ul>
    <footer>
      <PaginationBoxView
        activeClassName="active"
        breakLabel="..."
        containerClassName="pagination "
        disableInitialCallback={false}
        disabledClassName="disabled"
        marginPagesDisplayed={2}
        nextClassName="next"
        nextLabel={
          <i
            className="material-icons"
          >
            chevron_right
          </i>
        }
        onPageChange={[Function]}
        pageClassName="waves-effect"
        pageCount={24}
        pageRangeDisplayed={4}
        previousClassName="previous"
        previousLabel={
          <i
            className="material-icons"
          >
            chevron_left
          </i>
        }
      />
    </footer>
  </div>
</Connect(ManageRecipes)>
`;

exports[`Favorite Recipes Page should render a preloader 1`] = `
<Connect(ManageRecipes)>
  <div
    className="container page-manage__recipes"
  >
    <Preloader />
  </div>
</Connect(ManageRecipes)>
`;

exports[`Favorite Recipes Page should render with right amount of elements 1`] = `
<Connect(ManageRecipes)>
  <div
    className="container page-manage__recipes"
  >
    <div
      className="no-favorites"
    >
      <span>
        Welcome to your favorites!
      </span>
      <span>
        Tap 
        <i
          className="mdi mdi-heart"
        />
         to save any recipe you like,
      </span>
      <span>
        and it'll show up here.
      </span>
    </div>
  </div>
</Connect(ManageRecipes)>
`;
